microcosm_postgres/migrate.py

```diff
@@ -0,0 +1,203 @@
+    """
+    Generate the template for new migrations.
+
+    This function takes the place of the `script.py.mako` file in the alembic directory.
+
+    """
+    return dedent('''\
+    """
+    ${message}
+
+    Revision ID: ${up_revision}
+    Revises: ${down_revision | comma,n}
+    Create Date: ${create_date}
+
+    """
+    from alembic import op
+    import sqlalchemy as sa
+    ${imports if imports else ""}
+
+    # revision identifiers, used by Alembic.
+    revision = ${repr(up_revision)}
+    down_revision = ${repr(down_revision)}
+    branch_labels = ${repr(branch_labels)}
+    depends_on = ${repr(depends_on)}
+
+
+    def upgrade():
+        ${upgrades if upgrades else "pass"}
+
+
+    def downgrade():
+        ${downgrades if downgrades else "pass"}
+    ''')
+
+
+@contextmanager
+def patch_script_directory(graph):
+    """
+    Monkey patch the `ScriptDirectory` class, working around configuration assumptions.
+
+    Changes include:
+      - Using a generated, temporary directory (with a generated, temporary `script.py.mako`)
+        instead of the assumed script directory.
+      - Using our `make_script_directory` function instead of the default `ScriptDirectory.from_config`.
+      - Using our `run_online_migration` function instead of the default `ScriptDirectory.run_env`.
+      - Injecting the current object graph.
+    """
+    temporary_dir = mkdtemp()
+    from_config_original = getattr(ScriptDirectory, "from_config")
+    run_env_original = getattr(ScriptDirectory, "run_env")
+
+    # use a temporary directory for the revision template
+    with open(join(temporary_dir, "script.py.mako"), "w") as file_:
+        file_.write(make_script_py_mako())
+        file_.flush()
+
+    # monkey patch our script directory and migration logic
+    setattr(ScriptDirectory, "from_config", classmethod(make_script_directory))
+    setattr(ScriptDirectory, "run_env", run_online_migration)
+    setattr(ScriptDirectory, "graph", graph)
+    try:
+        yield temporary_dir
+    finally:
+        # cleanup
+        delattr(ScriptDirectory, "graph")
+        setattr(ScriptDirectory, "run_env", run_env_original)
+        setattr(ScriptDirectory, "from_config", from_config_original)
+        rmtree(temporary_dir)
+
+
+def main(graph, *args):
+    """
+    Entry point for invoking Alembic's `CommandLine`.
+
+    Alembic's CLI defines its own argument parsing and command invocation; we want
+    to use these directly but define configuration our own way. This function takes
+    the behavior of `CommandLine.main()` and reinterprets it with our patching.
+
+    :param graph: an initialized object graph
+    :param migration_dir: the path to the migrations directory
+
+    """
+    if not hasattr(graph, "migrations_dir"):
+        raise Exception("Migrations dir must be defined")
+    if not isdir(graph.migrations_dir):
+        raise Exception("Migrations dir must exist: {}".format(graph.migrations_dir))
+
+    cli = CommandLine()
```

Comment:
I'm directly invoking Alembic's CLI. Note that `Flask-Migrate` actually redefines its own CLI argument parser for all of the Alembic commands (just so it fits in the general `Flask-Script` structure). I didn't think any of that was worthwhile.